### ISSUE TITLE
False positives in reusable workflows for org level secrets

### ISSUE THREAD

**Issue by bewuethr** [rhysd/actionlint#141]
I have a workflow that calls a reusable workflow and passes a secret:

```yaml
name: Caller workflow

on:
  workflow_dispatch:

jobs:
  call:
    name: Call reusable workflow
    uses: myorg/actions/.github/workflows/reusable.yml@main
    secrets:
      token: ${{ secrets.SPECIAL_TOKEN }}
```

In the reusable workflow, I refer to that secret, but I also use an org-level secret:

```yaml
name: Reusable workflow

on:
  workflow_call:
    secrets:
      token:
        description: A secret token
        required: true

jobs:
  something:
    name: Do something with secrets
    runs-on: ubuntu-20.04

    steps:
      - name: Print secrets
        run: |
          echo "${{ secrets.token }}"
          echo "${{ secrets.ORG_LEVEL_SECRET }}"
```

This job runs fine. However, actionlint complains:

```txt
reusable.yml:17:52: property "org_level_secret" is not defined in object type {actions_runner_debug: string; token: string; github_token: string; actions_step_debug: string} [expression]
```

I believe this is a false positive. The check seems to be described [here](https://github.com/rhysd/actionlint/blob/main/docs/checks.md#check-types-of-inputs-and-secrets-in-reusable-workflow). The assumption that every reference to the `secret` object must be declared in `on.workflow_call.secrets` doesn't hold in this case.

**Comment by rhysd** [comment]
@bewuethr That's because you didn't declare the `ORG_LEVEL_SECRET` secret. Is it set by GitHub Actions runtime? If so, please provide the document of the secret. Otherwise, you need to declare it:

```yaml
on:
  workflow_call:
    secrets:
      token:
        description: A secret token
        required: true
      org_level_secret:
        description: ...

**Comment by rhysd** [comment]
OK, I understood your point. The secret is set in org-level. So it is available in workflows of the org even if it is not declared. Am I correct?

Hmm, this is difficult for actionlint because actionlint does not know what secrets are set in org-level.

**Comment by bewuethr** [comment]
> OK, I understood your point. The secret is set in org-level. So it is available in workflows across the org even if it is not declared. Am I correct?
> 
> Hmm, this is difficult for actionlint because actionlint does not know what secrets are set in org-level.

Yes, this is exactly right. It's a secret available to every repository in the organization. I can just pass such a secret to the reusable workflow, but I'd like not to because that keeps the call site simpler (my actual workflow uses half a dozen such secrets).

Maybe something like a default setting for a secret?

```yaml
on:
  workflow_call:
    secrets:
      org_level_secret:
        description: An org level secret
        required: false
        default: ${{ secrets.ORG_LEVEL_SECRET }}
```

so at least within the workflow you'd only reference known secrets?

**Comment by faubion-hbo** [comment]
@bewuethr i don't believe your call to `${{ secrets.ORG_LVL_SECRET }}` from within a reusable workflow would return the secrets and will instead result in an empty string. it wasn't until recently that `secrets: inherit` was released, and from [this blog post about that](https://github.blog/changelog/2022-05-03-github-actions-simplify-using-secrets-with-reusable-workflows/) it sounds like it was never supported to access organization secrets from within an reusable workflow.

i've tested this myself and would only get empty strings when the repo containing the workflow (and the caller workflow's repo) tried to get the organization secret directly.

**Comment by bewuethr** [comment]
> @bewuethr i don't believe your call to `${{ secrets.ORG_LVL_SECRET }}` from within a reusable workflow would return the secrets and will instead result in an empty string. it wasn't until recently that `secrets: inherit` was released, and from [this blog post about that](https://github.blog/changelog/2022-05-03-github-actions-simplify-using-secrets-with-reusable-workflows/) it sounds like it was never supported to access organization secrets from within an reusable workflow.
> 
> i've tested this myself and would only get empty strings when the repo containing the workflow (and the caller workflow's repo) tried to get the organization secret directly.

Oh, you're totally right! My workflow failed silently without the secret, but it really wasn't available. I can confirm that an org-level secret isn't available in a reusable workflow without `secrets: inherit`. I somehow assumed it would be, as the reusable workflow is run in the context of the caller workflow.

I'll close this then.

I guess I can't do my "use defaults for some secrets and pass others explicitly" approach then.

---

Summary for anybody bumping into this issue:

- A reusable workflow has either access to the secrets defined in `workflow_call.secrets`, or to all the secrets the caller workflow has access to if it calls the reusable workflow with `secrets: inherit`
- Org-level secrets are *not* available in addition to what is available as per the first bullet point
- But, on the upside, actionlint warns about doing it wrong 😉
